ledger/src/blockstore_meta.rs

```diff
@@ -11,6 +12,24 @@ use {
     },
 };
 
+bitflags! {
+    #[derive(Default, Serialize, Deserialize)]
+    /// Flags to indicate whether a slot is a descendant of a sot on the main fork
+    pub struct ConnectedFlags:u8 {
+        // A slot can be connected in two scenarios:
+        // 1) If the parent is connected and if the slot is full. By this
+        //    definition alone, the node would need to have blocks all the way
+        //    way back to genesis that are connected.
+        // 2) If ledger processing began at this slot from a snapshot. Snapshots
+        //    are presumed roots, and as such, we extend the idea of being
+        //    connected to this slot so that nodes starting from snapshot can
+        //    build a connected chain.
+        const CONNECTED        = 0b0000_0001;
+        /// Whether this
+        const PARENT_CONNECTED = 0b1000_0000;
```

Comment:
The flag design here seems to allow PARENT_CONNECTED to be false and CONNECTED to be true (which should never be the case).  How about something like we have IS_FULL and PARENT_CONNECTED, and we define CONNECTED to be IS_FULL & PARENT_CONNECTED?

Reply:
> The flag design here seems to allow PARENT_CONNECTED to be false and CONNECTED to be true (which should never be the case).

So that is exactly the reason why we need the second bit (`is_connected` and `is_parent_connected`). When we load from a snapshot at slot `S`, we don't actually need the shreds for that slot since the loaded bank is the state after transactions have been replayed at `S`. Moreso, we won't even try to repair this slot if we're starting from a clean ledger, so that slot will explicitly not be full unless there is manual intervention (a user copies shreds from somewhere).

This slot is not full, so we can't rely on `is_parent_connected && is_full()` as the criteria for whether `S` is connected or not. Thus, we introduce the second bit

Reply:
Let me see if I understand this.

So when we load from a snapshot at slot `S`, did the snapshot already include the complete replay of slot `S`?
If the answer is yes (and I think the answer should be yes), then slot `S` is connected (and at least was full before, otherwise the snapshot won't be able to include slot `S`).

So for slot `S`, can I know where we mark it as connected?  In addition, if we simply mark slot `S` as parent-connected and full, can I know what's the difference between marking `S` as connected (but not full)?

Because it seems to me that only the most recent slot in the snapshot could have such a special case, but the overall bit design could potential allow every slot to have such a special case.

Reply:
> So when we load from a snapshot at slot S, did the snapshot already include the complete replay of slot S

Yes, a snapshot includes the frozen state that includes slot `S`'s transactions

> then slot S is connected (and at least was full before, otherwise the snapshot won't be able to include slot S

- This is true when considering a node that is loading from its' own snapshot.
- When a node has no local state (joining for the first time or joining with a wiped ledger), they have to get their snapshot from someone else. They download a snapshot at slot `S`, and unpacking / generating things re-establishes their bank/account state with `S` included / ready to play `S`'s child. Since the bank already has `S` reflected in it, they don't pull down the block (shreds) into rocksdb. As such, `S` will not be full in the blockstore on this node that just joined.

Building on the download snapshot case, we mark `S` connected. Now suppose that `T` is `S`'s child. When we download the shreds for `T`, we insert a SlotMeta for `T`. When inserting this, we check `S`'s SlotMeta, see that `S` is connected, and then mark `T`  as having its' parent connected. Then, when `T` gets the last shred it needs to become full, we update `T` to be connected, and then propagate the fact that `T` is connected to `T`'s children (by marking `T`'s children with the `PARENT_CONNECTED` bit).

> In addition, if we simply mark slot S as parent-connected and full, can I know what's the difference between marking S as connected (but not full)?

Fullness is reflective of if we have received all the shreds for a slot (the last shred in the block has a marker so we can know when we shouldn't expect anymore). Using the same example as above where `S` is a snapshot slot that we downloaded from the cluster, we can mark it is parent-connected, but marking it "full" would be inconsistent since we don't have the shreds.

That is why I left the `is_full()` to be determined based on what shreds are present, and called the bits `parent_connected` && `connected`

Reply:
Thanks for the extra information.  It makes sense and sounds good.  Let's make it clear in the code comment about the definition of CONNECTED.  CONNECTED means:

* The slot has been a root (and this implicitly implies its parent slot is connected), or
* Its parent slot is connected and the slot itself is full.

So it means CONNECTED should also guarantee and cover PARENT_CONNECTED.

I think we should define the bits like the following:

* PARENT_CONNECTED: its parent is connected.
* FULLY_REPLAYABLE (or simply call REPLAYABLE): the Slot is fully replayable or has been fully replayed.
* CONNECTED:  PARENT_CONNECTED & FULLY_REPLAYABLE.

Reply:
> So it means CONNECTED should also guarantee and cover PARENT_CONNECTED.

For any slot except a snapshot slot, yes. But again, the snapshot slot is the special case where we are saying that the slot is connected, despite its' parent not existing in the blockstore and us not actually having the shreds.

In regards to your suggestion of adding an extra bit, I disagree. In this case, `FULLY_REPLAYABLE` simply means that the slot is full. We already track this with other fields in the `SlotMeta`, so adding another bit that we'd have to update / keep aligned adds extra burden

Reply:
The reason behind my suggestion is that I still feel it's quite odd to allow CONNECTED to be true but PARENT_CONNECTED to be false.  I understand the exception here are genesis and snapshot slot, but I feel allowing such a combination seems error-prone to me, and we also need to add assertions at proper places to make sure regular slots never have CONNECTED to be true but PARENT_CONNECTED to be false.

> In regards to your suggestion of adding an extra bit

I think you probably misunderstood my suggestion.  It is still two bits, but CONNECTED means both PARENT_CONNECTED and FULLY_REPLAYABLE are true.  So no extra bit for CONNECTED.  Fully replayable is not the same as full.  For instance, genesis slot and snapshot slots are both replayable as well.

Reply:
> For instance, genesis slot and snapshot slots are both replayable as well.

We recreate the bank from the snapshot and do not have the slot; thus, this slot is not replayable

Reply:
Maybe we can pick a different name for that bit.  Originally I was thinking about REPLAYABLED_OR_REPLAYED, but I think given that a slot is already in a snapshot or is a genesis slot, it means the slot has already replayed in the past, so it was replayable before thus I simplified it to be REPLAYABLE.

I still feel using `11` to represent CONNECTED is better and less error-prone than allowing `10` and `11`.

Reply:
> I still feel using 11 to represent CONNECTED is better and less error-prone than allowing 10 and 11 to both represent CONNECTED.

Thanks for stating this - I now fully understand your hangup here. Let me mull on this and then let's chat on Zoom to hash this out, should get to a resolution quicker that way